utils/workspace.js

```diff
@@ -0,0 +1,205 @@
+#!/usr/bin/env node
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+//@ts-check
+
+/**
+ * Use the following command to typescheck this file:
+ * npx tsc --target es2020  --watch --checkjs --noemit --moduleResolution node workspace.js
+ */
+const fs = require('fs');
+const path = require('path');
+const util = require('util');
+const url = require('url');
+
+const readJSON = async (filePath) => JSON.parse(await fs.promises.readFile(filePath, 'utf8'));
+const writeJSON = async (filePath, json) => {
+  await fs.promises.writeFile(filePath, JSON.stringify(json, null, 2) + '\n');
+}
+
+class PWPackage {
+  constructor(descriptor) {
+    this.name = descriptor.name;
+    this.path = descriptor.path;
+    this.files = descriptor.files;
+    this.packageJSONPath = path.join(this.path, 'package.json');
+    this.packageJSON = JSON.parse(fs.readFileSync(this.packageJSONPath, 'utf8'));
+    this.isPrivate = !!this.packageJSON.private;
+  }
+}
+
+class Workspace {
+  /**
+   * @param {string} rootDir
+   * @param {PWPackage[]} packages
+   */
+  constructor(rootDir, packages) {
+    this._rootDir = rootDir;
+    this._packages = packages;
+  }
+
+  packages() { return this._packages; }
+
+  /**
+   * @param {string} version
+   */
+  async setVersion(version) {
```

Comment:
Let's preserve `Version must not start with "v"` error here, because some other script calls it.

Reply:
done